### ISSUE TITLE
compilation API does not check racket version

### ISSUE THREAD

**Issue by Metaxal** [racket/racket#3480]
**What version of Racket are you using?**
7.9 cs, but appears with (likely) all version of racket.

**How to reproduce the error**

Write a short module file "afile.rkt" such as:
```racket
#lang racket
(displayln "Hello")
```
Compile it with a *older* version of `raco`, say 7.8.

Now, attempt to recompile the same file using the API, say `try-compile.rkt`:
```racket
#lang racket/base
(require compiler/cm)
(managed-compile-zo "...afile.rkt")
```
and run this module with the newer racket, say 7.9: `racket try-compile.rkt`.
There's no output, no error message (so supposedly it goes fine).

Run `racket afile.rkt` with the newer racket (7.9), this fails with:
```
read-compiled-linklet: version mismatch  expected: "7.9.0.3"  found: "7.8.0.6"  in: .../afile_rkt.zo
```

**What should have happened?**
`managed-compiled-zo` and friends should check the racket version and recompile if there's a mismatch.

**Please include any other relevant details**

The issue is that the racket version check appears only here:
https://github.com/racket/racket/blob/8b917039a28b32316bed537813d3dd08198f798c/racket/src/expander/compile/read-linklet.rkt#L10
and is not checked by the compilation API.

I don't know why `raco` behaves differently though.

Context: This is for quickscript, which automatically recompiles the user scripts. Such an error appears when the user installs a new version of racket. Since the user scripts are not dependencies of drracket, `raco setup` doesn't recompile them.

cc @rfindler

**Comment by mflatt** [comment]
I'm not able to replicate this problem, so far. Can you provide more explicit step-by-step instructions, showing the exact commands you ran on the command and/or in [Dr]Racket?

**Comment by Metaxal** [comment]
Uh, my apologies. I just ran the procedure I'm describing above and it does work fine.

I took a shortcut from my setup where the issue happens. The difference is that I was running the compilation from the quickscript tool/plugin (since that's the use case), so inside DrRacket. 

Could it be that there are a few parameters that need to be set properly from within a DrRacket tool so as to emulate what happens on the command line for `managed-compile-zo` and friends?

If nothing comes to mind, I'll try to write a minimal plugin to reproduce the error systematically.

**Comment by rfindler** [comment]
Perhaps the namespace? If a file is already loaded into the current
namespace then it won't be looked at to recompile?

Robby

**Comment by mflatt** [comment]
Definitely reusing a namespace would cause a problem, since `managed-compile-zo` works by loading modules. I'll update the documentation to explain that.

**Comment by Metaxal** [comment]
Thanks Matthew for updating the docs. Just to let you know, parameterizing
the current namespace with `(make-base-empty-namespace)` didn't solve the
problem.

**Comment by Metaxal** [comment]
Here's a (almost minimal) package that reproduces the compilation issue:
https://gist.github.com/Metaxal/616b2bc0ccdd05d2a423977f9a6408f1

(The file "tool.rkt" is irrelevant. It's a drracket plugin because initially I thought due to some interaction with DrRacket, but it appears this happens also on the command line.)

The main relevant file is `utils.rkt`, plus what the `scripts` subdirectory contains.

Install:
```shell
raco pkg install https://github.com/Metaxal/qs-compile-issue.git
```
Then run:
```shell
export PLTSTDERR=debug@trycompile && racket -l qs-compile-issue/utils -- --erase
```
The `--erase` flag makes the program remove bad zo files before compiling. You should see this correct output:
```
trycompile: Erasing bad zos before compiling
trycompile: File: /home/laurent/Prog/Racket/qs-compile-issue/scripts/test-compile.rkt
trycompile:   Compiled for (7.7.0.901 racket) (current is 7.9.0.3 chez-scheme)
trycompile:   Compiling.
trycompile:   Compiled for (7.9.0.3 chez-scheme) (after compiling)
trycompile: File: /home/laurent/Prog/Racket/qs-compile-issue/scripts/test-compile-cs.rkt
trycompile:   Compiled for (7.8.0.6 chez-scheme) (current is 7.9.0.3 chez-scheme)
trycompile:   Compiling.
trycompile:   Compiled for (7.9.0.3 chez-scheme) (after compiling)
```
Observe how the version number has changed for each file. Now run this (no `--erase` flag):
```
export PLTSTDERR=debug@trycompile && racket -l qs-compile-issue/utils
```
You should see this incorrect output:
```
trycompile: File: /home/laurent/Prog/Racket/qs-compile-issue/scripts/test-compile.rkt
trycompile:   Compiled for (7.7.0.901 racket) (current is 7.9.0.3 chez-scheme)
trycompile:   Compiling.
trycompile:   Compiled for (7.7.0.901 racket) (after compiling)
trycompile: File: /home/laurent/Prog/Racket/qs-compile-issue/scripts/test-compile-cs.rkt
trycompile:   Compiled for (7.8.0.6 chez-scheme) (current is 7.9.0.3 chez-scheme)
trycompile:   Compiling.
trycompile:   Compiled for (7.8.0.6 chez-scheme) (after compiling)
```
What this means:
First, 2 old compilation files from previous racket versions are copied to the `compiled` directory. Their timestamps are thus newer than the corresponding `.rkt` files.

**Comment by Metaxal** [comment]
It can also be checked that the files have not been compiled properly with:
```
$ racket -l qs-compile-issue/scripts/test-compile
read-compiled-linklet: version mismatch  expected: "7.9.0.3"  found: "7.7.0.901"  in: /home/laurent/Prog/Racket/qs-compile-issue/scripts/compiled/test-compile_rkt.zo
```

**Comment by mflatt** [comment]
I think you forgot to commit a "test-compile.rkt" source, but it was easy enough to add one.

For version checking, the compilation manager relies on ".dep" files alongside ".zo" files. So, the test here should copy an old ".dep" file into place as well as an old ".zo" file. Indeed, the compilation manager doesn't look at the version inside a ".zo" file, and it instead requires that ".dep" files are consistent (when they are present) with ".zo" files.

When I make those adjustments, then things seem to work without `--erase`.

Meanwhile, now that I understand you're using `make-caching-managed-compile-zo`, I see that the documentation is lacking in another way: that function needs to be called with the new namespace in place, because it remembers the namespace at the time that the caching manager is created. More generally, the documentation should clarify that `make-caching-managed-compile-zo` creates a handler with `make-compilation-manager-load/use-compiled-handler`, which is the part that depends on the current namespace.

**Comment by Metaxal** [comment]
I see. I'll add dep files to my tests.

Shouldn't the compiler discard the zo file if no dep file is found? It seems to be blocking compilation instead.

**Comment by mflatt** [comment]
Yes, when a ".dep" file is missing, the compilation manager assumes the worst and recompiles. That's what happens for me. My best guess is that an old ".dep" was still there in your run (e.g., created by a run with `--erase`).

**Comment by Metaxal** [comment]
Yes, that makes sense. Thank you for you answers on a Sunday!

**Comment by Metaxal** [comment]
The `quickscript-test` test cases now pass with all this.

**Comment by rfindler** [comment]
I don't think that the characterization of when the namespace, etc is looked up is quite right. The implementation of [`make-caching-managed-compile-zo`](https://github.com/racket/racket/blob/master/racket/collects/compiler/private/cm-minimal.rkt#L99-L102) seems to suggest that none of the parameters are consulted when it is first called. Instead, they are consulted when the function it returns is first given a path. The gist above seems to be gone now but the code that @Metaxal showed me in slack had the namespace parameter set at the correct time (it was after the call to `make-caching-managed-compile-zo` but before the first file was supplied to the function that `make-caching-managed-compile-zo` returned).

**Comment by mflatt** [comment]
Thanks for the correction, and I'll update the docs. The handler is created every time the result from `make-caching-managed-compile-zo` is called, right?

**Comment by Metaxal** [comment]
(sorry, the gist wasn't supposed to be posted—I turned it into a package afterwards.)

**Comment by rfindler** [comment]
@Metaxal : what was the fix that you found? I'm still not clear on why the original code was wrong, actually (the code in the repo linked to above).

**Comment by Metaxal** [comment]
For the record, the bug was in the test: I needed to copy also the `.dep` files for the compiler to see that the source file was compiled with a different version of racket. Now all quickscript-test tests pass, and drdr doesn't seem to complain.

@rfindler I'll let you close if you're satisfied with this.

**Comment by rfindler** [comment]
Thanks!
